### DISCUSSION TITLE
The dreaded "Deprecated" and "Obsolete"

### DISCUSSION THREAD

**Discussion by thabaum** [dnnsoftware/Dnn.Platform#4091] [General] [answered]
Looking at the SMTP controllers and noticed old things needing to get replaced with the new. 

![image](https://user-images.githubusercontent.com/13577556/93356766-7548af00-f7f4-11ea-9971-60cdda65dbe1.png)

I was trying to utilize some of what it recommends.  I am wondering how I may be able to help cleanup some of these things.  I know @ahoefling had recommended treating the warnings as errors.  

Wouldn't this be one of them here?  And if so I wish to learn how I can help resolve these issues for the platform.

**Comment by bdukes** [comment]
@thabaum one thing to be aware of with SMTP specifically is that there's a pending pull request #3969 to adjust a lot of that code and move it into a provider.  So I check to make sure whatever you're working with isn't about to be changed anyway (you should be good to adjust the Persona Bar pieces, but the core pieces will be shifting around).

**Reply by thabaum** [reply]
This was just an example, I am aware of the SMTP changes.  I was not going to work on this until I get a full understanding of any new changes going into 10.  I am warming up for 10 being busy with this type of work replacing things...

**Comment by bdukes** [comment, marked as answer]
For these deprecations where we're moving from `Whatever.Instance` to `IWhatever`, to resolve the warning you need to start retrieving the interface via dependency injection.  Exactly how you do that will depend a lot on which part of the code you're in.  If you're looking at a Web API controller (e.g. in the Persona Bar code), you can just add an additional constructor dependency and DI will fill it in automatically. For instance,

```c#
public class ServerSettingsSmtpHostController : PersonaBarApiController
{
    private readonly IHostSettingsService hostSettingsService;

    public ServerSettingsSmtpHostController(IHostSettingsService hostSettingsService)
    {
        this.hostSettingsService = hostSettingsService;
    }
}
```

You can then replace `HostController.Instance.GetString` with `this.hostSettingsService.GetString`.

**Reply by SkyeHoefling** [reply]
@bdukes you beat me to it, this is exactly what I was going to recommend. 

Many of these deprecations require the usage of Dependency Injection. @thabaum if you are working in an area of DNN that does not support Dependency Injection I urge you to create an issue on the backlog and document where we need to add it in DNN. The goal is to make Dependency Injection accessible in every part of the platform where possible. We are currently a long way off from that goal, but the more people documenting the lack of the support the more likely we can get it done faster 💯

**Comment by thabaum** [comment]
Great thanks, I think that clears it up a lot.

These marked for obsolete I presume they all have the `IWhatever` ready to go I just need to use DI to bring it in to use?

I will give it a go in an area that is not worked on much first...

**Reply by bdukes** [reply]
Yes.  I also want to note that with that specific deprecation, some of those messages point to the wrong type (it got renamed after those messages were started, fix is #4092)

**Reply by SkyeHoefling** [reply]
There is a lot of work happening in Dependency Injection space right now and as we move more and more core components to the abstractions library we are finding better conventions to follow. We are doing our best to keep the deprecation notices up-to-date. 

If you find anything that seems out of place, I encourage you to start a discussion and we can then decide if we need to create an item on the backlog to fix it.

**Reply by thabaum** [reply]
I will help when I can as this helps make things much clearer what is happening currently behind the scenes in DNN with respect to DI.

Thank you @bdukes and @ahoefling for getting me up to speed.  I will watch future related PR's and dig into a few of the past to help connect the dots further.  If you have a great example of of a PR adding DI where needed please share I would love reviewing it!

**Reply by SkyeHoefling** [reply]
You can take a look at the PRs I have been submitting recently
- #4052 
- #4021 
- #3990 
- #3988 

Those are all completed PRs, there is a lot of dependency injection work going on right now so plenty of examples to share.
